test/e2e/prometheusadapter_test.go

```diff
@@ -25,10 +25,138 @@ import (
+			return true
+		}
+	}
+	return false
+}
+
+func isAPIServiceAvailable(conditions []apiservicesv1.APIServiceCondition) bool {
+	for _, condition := range conditions {
+		if condition.Type == apiservicesv1.Available && condition.Status == apiservicesv1.ConditionTrue {
+			return true
+		}
+	}
+	return false
+}
+
+func TestMetricsAPIAvailability(t *testing.T) {
+	var lastErr error
+	err := wait.Poll(time.Second, 5*time.Minute, func() (bool, error) {
+		metricsService, err := f.APIServicesClient.ApiregistrationV1().APIServices().Get("v1beta1.metrics.k8s.io", metav1.GetOptions{})
+		lastErr = errors.Wrap(err, "getting metrics APIService failed")
+		if err != nil {
+			return false, nil
+		}
+		if !isAPIServiceAvailable(metricsService.Status.Conditions) {
+			lastErr = errors.New("v1beta1.metrics.k8s.io apiservice is not available")
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		if err == wait.ErrWaitTimeout && lastErr != nil {
+			err = lastErr
+		}
+		t.Fatal(err)
+	}
+}
+
+func TestNodeMetricsPresence(t *testing.T) {
+	var lastErr error
+	err := wait.Poll(time.Second, 5*time.Minute, func() (bool, error) {
+		nodes, err := f.KubeClient.CoreV1().Nodes().List(metav1.ListOptions{})
+		lastErr = errors.Wrap(err, "getting nodes list failed")
+		if err != nil {
+			return false, nil
+		}
+		nodeMetrics, err := f.MetricsClient.MetricsV1beta1().NodeMetricses().List(metav1.ListOptions{})
+		lastErr = errors.Wrap(err, "getting metrics list failed")
+		if err != nil {
+			return false, nil
+		}
+		if len(nodes.Items) != len(nodeMetrics.Items) {
+			lastErr = errors.New("number of nodes doesn't match number of node metrics reported")
+			return false, nil
+		}
+		for _, item := range nodeMetrics.Items {
+			if !isNodeInNodesList(item.Name, nodes.Items) {
+				lastErr = errors.New("node reporting metrics couldn't be found in nodes list")
+				return false, nil
+			}
+			if item.Usage.Cpu() == nil || item.Usage.Memory() == nil {
+				lastErr = errors.New("node cpu or memory metric not found")
+				return false, nil
+			}
+		}
+		return true, nil
+	})
+	if err != nil {
+		if err == wait.ErrWaitTimeout && lastErr != nil {
+			err = lastErr
+		}
+		t.Fatal(err)
+	}
+}
+
+func TestPodMetricsPresence(t *testing.T) {
+	var lastErr error
+	err := wait.Poll(time.Second, 5*time.Minute, func() (bool, error) {
+		pods, err := f.KubeClient.CoreV1().Pods("").List(metav1.ListOptions{})
+		lastErr = errors.Wrap(err, "getting pods list failed")
+		if err != nil {
+			return false, nil
+		}
+		podMetrics, err := f.MetricsClient.MetricsV1beta1().PodMetricses("").List(metav1.ListOptions{})
+		lastErr = errors.Wrap(err, "getting metrics list failed")
+		if err != nil {
+			return false, nil
+		}
+		if len(pods.Items) != len(podMetrics.Items) {
+			lastErr = errors.New("number of pods doesn't match number of pod metrics reported")
+			return false, nil
+		}
+
+		for _, pod := range podMetrics.Items {
+			if !isPodInPodsList(pod.Name, pods.Items) {
```

Comment:
a bit of a nit, but this should be the combination of pod and namespace, I realize this is highly unlikely, but let's not leave the possibility open

Reply:
Added with a workaround for all namespaces (when namespace == "").

Reply:
I don't think that'll work. That should be `pod.Namespace`. Using `""` is just for _selecting_ all metrics, the calls do return pods with their respective namespace set. What I meant with the above comment was we should make sure the pod name + namespace combination is checked for as there can be pods with the same name in different namespaces.

Reply:
Sorry, I misunderstood you. Now it should be ok :)